File: src/java/org/voegtle/weatherstation/server/logic/WeatherDataImporter.java
package org.voegtle.weatherstation.server.logic;

import org.voegtle.weatherstation.server.parser.DataLine;
import org.voegtle.weatherstation.server.parser.DataParser;
import org.voegtle.weatherstation.server.persistence.*;
import org.voegtle.weatherstation.server.request.ResponseCode;
import org.voegtle.weatherstation.server.util.DateUtil;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class WeatherDataImporter {
  private static final Logger log = Logger.getLogger(WeatherDataImporter.class.getName());

  private final PersistenceManager pm;
  private final DateUtil dateUtil;
  private final Date importedUntil;
  private final DataIndicies dataIndicies;
  private LocationProperties locationProperties;

  public WeatherDataImporter(PersistenceManager pm, LocationProperties locationProperties) {
    this.pm = pm;
    this.locationProperties = locationProperties;
    this.dateUtil = locationProperties.getDateUtil();
    this.importedUntil = getDateOfLastDataSet();
    dataIndicies = locationProperties.getDataIndices();
  }

  public String doImport(ArrayList<DataLine> lines) {
    String result;
    try {
      boolean persisted = false;
      DataParser parser = new DataParser(dateUtil, dataIndicies);
      List<WeatherDataSet> dataSets = parser.parse(lines);
      log.info("Number of valid datasets: " + dataSets.size());
      for (WeatherDataSet dataSet : dataSets) {
        if (isNotOutdated(dataSet)) {
          if (pm.makePersitant(dataSet)) {
            persisted = true;
          }
        }
      }

      if (persisted) {
        new WeatherDataSmoother(pm, dateUtil).smoothWeatherData();
        new WeatherDataAggregator(pm, dateUtil).aggregateWeatherData();
        new WeatherDataForwarder(pm, locationProperties).forwardLastDataset();
      } else {
        log.warning("no dataset has been persisted");
      }

      result = persisted ? ResponseCode.ACKNOWLEDGE : ResponseCode.IGNORED;
    } catch (ParseException ex) {
      log.log(Level.SEVERE, "parsing failed", ex);
      result = ResponseCode.PARSE_ERROR;
    }
    return result;
  }

  private boolean isNotOutdated(WeatherDataSet dataSet) {
    boolean notOutdated = importedUntil.before(dataSet.getTimestamp());
    if (!notOutdated) {
      log.warning("WeatherDataSet from " + dataSet.getTimestamp() + " is outdated. import until: " + importedUntil);
    }
    return notOutdated;
  }

  private Date getDateOfLastDataSet() {
    Date lastImport = dateUtil.getDate(2016, 1, 1);

    WeatherDataSet youngestDataSet = pm.fetchYoungestDataSet();
    if (youngestDataSet != null && youngestDataSet.getTimestamp().after(lastImport)) {
      lastImport = youngestDataSet.getTimestamp();
    }

    SmoothedWeatherDataSet youngestSmoothedDS = pm.fetchYoungestSmoothedDataSet();
    if (youngestSmoothedDS != null && youngestSmoothedDS.getTimestamp().after(lastImport)) {
      lastImport = youngestSmoothedDS.getTimestamp();
    }
    return lastImport;
  }

}

File: src/main/java/org/voegtle/weatherstation/server/logic/WeatherDataImporter.kt
package org.voegtle.weatherstation.server.logic

import org.voegtle.weatherstation.server.logic.caching.WeatherDataProvider
import org.voegtle.weatherstation.server.parser.DataLine
import org.voegtle.weatherstation.server.parser.DataParser
import org.voegtle.weatherstation.server.persistence.PersistenceManager
import org.voegtle.weatherstation.server.persistence.entities.LocationProperties
import org.voegtle.weatherstation.server.persistence.entities.WeatherDataSet
import org.voegtle.weatherstation.server.request.ResponseCode
import java.text.ParseException
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger

class WeatherDataImporter(private val pm: PersistenceManager, private val locationProperties: LocationProperties) {
    private val log = Logger.getLogger(WeatherDataImporter::class.java.name)
    val weatherDataProvider = WeatherDataProvider(pm)

    private val dateUtil = locationProperties.dateUtil
    private val importedUntil = findDateOfLastDataSet()
    private val dataIndicies = locationProperties.dataIndices
    var persisted: Int = 0
        private set

    private fun findDateOfLastDataSet(): Date {
        var lastImport = dateUtil.getDate(2023, 10, 1)

        val youngestDataSet = weatherDataProvider.getYoungestWeatherDataSet()
        if (youngestDataSet.timestamp.after(lastImport)) {
            lastImport = youngestDataSet.timestamp
        }

        val youngestSmoothedDS = pm.fetchYoungestSmoothedDataSet()
        if (youngestSmoothedDS != null && youngestSmoothedDS.timestamp.after(lastImport)) {
            lastImport = youngestSmoothedDS.timestamp
        }
        return lastImport
    }

    fun doImport(lines: ArrayList<DataLine>): String {
        var result: String
        try {
            log.info("Number of lines: " + lines.size)
            val parser = DataParser(dateUtil, dataIndicies)
            val dataSets = parser.parse(lines)
            log.info("Number of parsed datasets: " + dataSets.size)
            dataSets
                .filter { isNotOutdated(it) }
                .forEach {
                    weatherDataProvider.write(it)
                    persisted++
                }

            if (persisted > 0) {
                WeatherDataSmoother(pm, dateUtil).smoothWeatherData()
                WeatherDataAggregator(pm, dateUtil).aggregateWeatherData()
                WeatherDataForwarder(pm, locationProperties).forwardLastDataset()
            } else {
                log.warning("no dataset has been persisted")
            }

            result = if (persisted > 0) ResponseCode.ACKNOWLEDGE else ResponseCode.IGNORED
        } catch (ex: ParseException) {
            log.log(Level.SEVERE, "parsing failed", ex)
            result = ResponseCode.PARSE_ERROR
        }

        return result
    }

    private fun isNotOutdated(dataSet: WeatherDataSet): Boolean {
        val notOutdated = importedUntil.before(dataSet.timestamp)
        if (!notOutdated) {
            log.warning("WeatherDataSet from " + dataSet.timestamp + " is outdated. import until: " + importedUntil)
        }
        return notOutdated
    }

}
